Realm/Realm/Native/SyncSocketProvider.EventLoop.cs

```diff
@@ -0,0 +1,144 @@
+                _cancellationToken = cancellationToken;
+            }
+
+            public void Execute()
+            {
+                var status = Status.OK;
+                if (_cancellationToken.IsCancellationRequested)
+                {
+                    status = new(ErrorCode.OperationAborted, "Timer canceled");
+                }
+
+                RunCallback(_nativeCallback, status);
+            }
+        }
+    }
+
+    private class EventLoopWork : IWork
+    {
+        private readonly IntPtr _nativeCallback;
+        private readonly Status _status;
+
+        // Belongs to SyncSocketProvider. When Native destroys the Provider we need to stop executing
+        // enqueued work, but we need to release all the callbacks we copied on the heap.
+        private readonly CancellationToken _cancellationToken;
+
+        public EventLoopWork(IntPtr nativeCallback, Status status, CancellationToken cancellationToken)
+        {
+            _nativeCallback = nativeCallback;
+            _status = status;
+            _cancellationToken = cancellationToken;
+        }
+
+        public void Execute()
+        {
+            if (_cancellationToken.IsCancellationRequested)
+            {
+                Logger.LogDefault(LogLevel.Trace, "Deleting EventLoopWork callback only because event loop was cancelled.");
+                NativeMethods.delete_callback(_nativeCallback);
+                return;
+            }
+
+            RunCallback(_nativeCallback, _status);
+        }
+    }
+
+    private static void RunCallback(IntPtr nativeCallback, Status status)
+    {
+        Logger.LogDefault(LogLevel.Trace, $"SyncSocketProvider running native callback {nativeCallback} with status {status.Code} \"{status.Reason}\".");
+
+        using var arena = new Arena();
+        NativeMethods.run_callback(nativeCallback, status.Code, StringValue.AllocateFrom(status.Reason, arena));
+    }
+
+    private async Task PostWorkAsync(IntPtr nativeCallback)
+    {
+        Logger.LogDefault(LogLevel.Trace, "Posting work to SyncSocketProvider event loop.");
+        await _workQueue.Writer.WriteAsync(new EventLoopWork(nativeCallback, Status.OK, _cts.Token));
+    }
+
+    private async partial Task WorkThread()
+    {
+        Logger.LogDefault(LogLevel.Trace, "Starting SyncSocketProvider event loop.");
+        try
+        {
+            while (await _workQueue.Reader.WaitToReadAsync())
+            {
+                while (_workQueue.Reader.TryRead(out var work))
+                {
+                    work.Execute();
+                }
+            }
+        }
+        catch (Exception e)
+        {
+            Logger.LogDefault(LogLevel.Error, $"Error occurred in SyncSocketProvider event loop {e.GetType().FullName}: {e.Message}");
+            if (!string.IsNullOrEmpty(e.StackTrace))
+            {
+                Logger.LogDefault(LogLevel.Trace, e.StackTrace);
+            }
+
+            throw;
```

Comment:
I don't think throwing in a background thread will bring down the process - you may get a warning or something, but in release, I think it's just ignored on most platforms.